src/app/cluster-building-blocks/QuieterReporting.h

```diff
@@ -172,13 +172,13 @@ class QuieterReportingAttribute
      * @return AttributeDirtyState::kMustReport if attribute must be marked dirty right away, or
      * AttributeDirtyState::kNoReportNeeded otherwise.
      */
-    AttributeDirtyState SetValue(const chip::app::DataModel::Nullable<T> & newValue, Timestamp now,
-                                 SufficientChangePredicate changedPredicate)
+    AttributeDirtyState SetValue(const DataModel::Nullable<T> & newValue, Timestamp now, SufficientChangePredicate changedPredicate)
     {
-        bool isChangeOfNull       = newValue.IsNull() ^ mValue.IsNull();
-        bool areBothValuesNonNull = !newValue.IsNull() && !mValue.IsNull();
+        bool isChangeOfNull         = newValue.IsNull() != mValue.IsNull();
+        bool areBothValuesNonNull   = !newValue.IsNull() && !mValue.IsNull();
+        bool areBothValuesDifferent = areBothValuesNonNull && (newValue.Value() != mValue.Value());
```

Comment:
Hmm.  I guess we don't want to just early-return if `newValue == mValue` because we want to always give `changedPredicate` a shot at marking dirty even if the value has not changed?

I think that's worth documenting here so that someone doesn't "optimize" away that capability.  We should also clearly document that change predicates need to be aware of the fact that they might be called even if the value is not actually changing.

That said, it seems odd to me to allow this.  Why would we ever want to mark things dirty if they are not actually changing?  It might help to include some of those use-cases as examples in the API documentation for change predicates or something....

Reply:
It's basically not to lose parity with MatterReportingAttributeChangeCallback() having been called unconditionally.

Reply:
Fixed